packages/manager/modules/iam/src/components/actionSelect/ActionTrees.class.js

```diff
@@ -0,0 +1,388 @@
+      ? selectedActions.filter(
+          ({ action: selectedAction }) =>
+            !actions.find(({ action }) => action === selectedAction),
+        )
+      : [];
+    this.buffer.customActionTree = actionTrees
+      ? actionTrees[0]
+      : {
+          actions: unknownActions
+            .sort(({ action: a }, { action: b }) => (a > b ? 1 : -1))
+            .map((action) => ({
+              resourceType: CUSTOM_RESOURCE_TYPE,
+              selected: true,
+              searchable: false,
+              value: action.action,
+            })),
+          expanded: false,
+          value: CUSTOM_RESOURCE_TYPE,
+        };
+  }
+
+  /**
+   * Displaying a lot of collapsibles containing a lot of checkboxes leads to performance issues
+   * We must limit the number of types to the ones selected by the user
+   * @param {Input} input
+   */
+  static limitByTypes(input) {
+    if (!input.resourceTypes?.length) {
+      throw new Error('NoTypesError');
+    }
+  }
+
+  /**
+   * Create an ActionTrees instance
+   * This instance is used to display collapsibles
+   * @param {Input} input
+   */
+  static createActionTrees(input) {
+    this.buffer.actionTrees = new ActionTrees(
+      ...this.buffer.actions
+        .map(({ resourceType: value }) => value)
+        .filter(Boolean)
+        .filter((value) => input.resourceTypes.includes(value))
+        .filter((value, index, list) => list.indexOf(value) === index)
+        .sort()
+        .map((value) => ({
+          expanded: false,
+          searchable: true,
+          searchQuery: input.actionTrees?.find(
+            (actionTree) => actionTree.value === value,
+          )?.searchQuery,
+          value,
+        })),
+    );
+  }
+
+  /**
+   * Each actionTree is given a list of categories
+   * These categories are extracted from the actions whose resourceType matches the actionTree's value
+   * @param {Input} input
+   */
+  static createCategories(input) {
+    this.buffer.actionTrees.forEach((actionTree) =>
+      Object.assign(actionTree, {
+        categories: input.actions
+          .filter((action) => action.resourceType === actionTree.value)
+          .filter(({ categories }) => categories?.length > 0)
+          .reduce((list, { categories }) => [...list, ...categories], [])
+          .filter((value, index, list) => list.indexOf(value) === index)
+          .sort()
+          .map((value) => ({
+            expanded: false,
+            selected: false,
+            value,
+          })),
+      }),
+    );
+  }
+
+  /**
+   * Each category of each actionTree is given a list of actions
+   * whose resourceType matches the actionTree's value
+   * @param {Input} input
+   */
+  static dispatchActions(input) {
+    this.buffer.actionTrees.forEachCategory(({ actionTree, category }) => {
+      const actions = input.actions
+        .filter((action) => action.resourceType === actionTree.value)
+        .filter((action) => action.categories.includes(category.value))
```

Comment:
**perf:** Both conditions can be combined to avoid looping twice.
```suggestion
        .filter((action) => action.resourceType === actionTree.value && action.categories.includes(category.value) )
```